Snowflake.Data/Logger/SecretDetector.cs

```diff
@@ -0,0 +1,113 @@
+        private static readonly string PRIVATE_KEY_DATA_PATTERN = @"""privateKeyData"": ""([a - z0 - 9 /+=\\n]{10,})""";
+        private static readonly string CONNECTION_TOKEN_PATTERN = @"(token|assertion content)([\'\""\s:=]+)([a-z0-9=/_\-\+]{8,})";
+        private static readonly string PASSWORD_PATTERN = @"(password|pwd)([\'\""\s:=]+)([a-z0-9!\""#\$%&\\\'\(\)\*\+\,-\./:;<=>\?\@\[\]\^_`\{\|\}~]{8,})";
+
+        private static string MaskAWSKeys(string text)
+        {
+            return Regex.Replace(text, AWS_KEY_PATTERN, @"$1='****'",
+                                         RegexOptions.IgnoreCase);
+        }
+
+        private static string MaskAWSTokens(string text)
+        {
+            return Regex.Replace(text, AWS_TOKEN_PATTERN, @"$1"":""XXXX""",
+                                         RegexOptions.IgnoreCase);
+        }
+
+        private static string MaskSASTokens(string text)
+        {
+            return Regex.Replace(text, SAS_TOKEN_PATTERN, @"$1=****",
+                                         RegexOptions.IgnoreCase);
+        }
+
+        private static string MaskPrivateKey(string text)
+        {
+            return Regex.Replace(text, PRIVATE_KEY_PATTERN, "-----BEGIN PRIVATE KEY-----\\\\nXXXX\\\\n-----END PRIVATE KEY-----",
+                                         RegexOptions.IgnoreCase | RegexOptions.Multiline);
+        }
+
+        private static string MaskPrivateKeyData(string text)
+        {
+            return Regex.Replace(text, PRIVATE_KEY_DATA_PATTERN, @"""privateKeyData"": ""XXXX""",
+                                         RegexOptions.IgnoreCase | RegexOptions.Multiline);
+        }
+
+        private static string MaskConnectionTokens(string text)
+        {
+            return Regex.Replace(text, CONNECTION_TOKEN_PATTERN, @"$1$2****",
+                                         RegexOptions.IgnoreCase);
+        }
+
+        private static string MaskPassword(string text)
+        {
+            return Regex.Replace(text, PASSWORD_PATTERN, @"$1$2****",
+                                         RegexOptions.IgnoreCase);
+        }
+
+        public static Mask MaskSecrets(string text)
+        {
+            Mask result = new Mask(maskedText: text);
+
+            if (String.IsNullOrEmpty(text))
+            {
+                return result;
+            }
+
+            try
+            {
+                result.maskedText =
+                    MaskConnectionTokens(
+                        MaskPassword(
+                            MaskPrivateKeyData(
+                                MaskPrivateKey(
+                                    MaskAWSTokens(
+                                        MaskSASTokens(
+                                            MaskAWSKeys(text)))))));
+
+                if (result.maskedText != text)
+                {
+                    result.isMasked = true;
+                }
+            }
+            catch (Exception ex)
```

Comment:
Can we add a log line in here to communicate that an exception occurred during secret masking?

Reply:
I tried this but there's an issue when both secret detector and logger call each other:
- Logger references SecretDetector to mask messages
- SecretDetector references Logger to log the exception

The logger could be changed from static to non-static but the logger was made non-static so that users can pass their own logger.

The exception thrown is recorded in `result.maskedText = ex.Message;` so that it shows on the log but let me know if you have suggestions